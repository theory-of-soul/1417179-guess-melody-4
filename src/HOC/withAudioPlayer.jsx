import React from "react";
import AudioPlayer from "../components/AudioPlayer/AudioPlayer";

const withAudioPlayer = (Component) => {
  return class WithAudioPlayer extends React.PureComponent {
    constructor(props) {
      super(props);
      this.state = {
        activePlayerId: 0
      };
    }

    render() {
      return (
        <Component
          {...this.props}
          renderAudioPlayer={(src, id) => {
            return (
              <AudioPlayer
                handlePlay={() => {
                  this.setState((prevProps) => ({
                    activePlayerId: prevProps.activePlayerId === id ? -1 : id
                  }));
                }}
                isPlaying={this.state.activePlayerId === id}
                src={src}
              />
            );
          }}
        />
      );
    }
  };
};

export default withAudioPlayer;
