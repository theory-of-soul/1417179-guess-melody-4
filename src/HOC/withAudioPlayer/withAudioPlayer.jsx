import React from "react";
import AudioPlayer from "../../components/AudioPlayer/AudioPlayer";
import withAudio from "../withAudio/withAudio";

const AudioPlayerWithAudio = withAudio(AudioPlayer);

const withAudioPlayer = (Component) => {
  return class WithAudioPlayer extends React.PureComponent {
    constructor(props) {
      super(props);
      this.state = {
        activePlayerId: 0
      };
    }

    render() {
      return (
        <Component
          {...this.props}
          renderAudioPlayer={(src, id) => {
            return (
              <AudioPlayerWithAudio
                handlePlay={() => {
                  this.setState((prevProps) => ({
                    activePlayerId: prevProps.activePlayerId === id ? -1 : id
                  }));
                }}
                isPlaying={this.state.activePlayerId === id}
                src={src}
              />
            );
          }}
        />
      );
    }
  };
};

export default withAudioPlayer;
