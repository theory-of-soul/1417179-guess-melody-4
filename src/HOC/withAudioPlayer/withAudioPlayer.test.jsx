import React from "react";
import renderer from 'react-test-renderer';
import withAudioPlayer from "./withAudioPlayer";
import PropTypes from "prop-types";

const src = ``;
const id = 0;
const options = {
  createNodeMock: (element) => {
    if (element.type === `audio`) {
      return {
        oncanplaythrough() {},
      };
    }
    return null;
  }
};

const MockComponent = (props) => {
  const {renderAudioPlayer} = props;

  return (
    <div>
      {renderAudioPlayer(src, id)}
    </div>
  );
};

MockComponent.propTypes = {
  renderAudioPlayer: PropTypes.func.isRequired,
};

const MockComponentWrapped = withAudioPlayer(MockComponent);

describe(`Hoc withAudioPlayer snapshot tests`, () => {
  it(`MockComponent with HOC`, () => {
    const component = renderer
      .create(<MockComponentWrapped/>, options)
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
