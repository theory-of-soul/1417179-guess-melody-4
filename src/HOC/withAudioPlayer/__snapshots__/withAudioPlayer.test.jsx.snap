// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hoc withAudioPlayer snapshot tests MockComponent with HOC 1`] = `
<div>
  <button
    className="track__button track__button--pause"
    disabled={true}
    onClick={[Function]}
    type="button"
  />
  <div
    className="track__status"
  >
    <audio
      src=""
    />
  </div>
</div>
`;
